lib/execution-profile.js

```diff
@@ -0,0 +1,129 @@
+"use strict";
+var utils = require('./utils');
+var types = require('./types');
+
+/**
+ * Creates a new instance of {@link ExecutionProfile}.
+ * @classdesc
+ * Represents a set configurations to be used in a statement execution.
+ * @param {Object} [options]
+ * @param {Number} [options.consistency]
+ * @param {LoadBalancingPolicy} [options.loadBalancing]
+ * @param {Number} [options.readTimeout]
+ * @param {RetryPolicy} [options.retry]
+ * @param {Number} [options.serialConsistency]
+ * @constructor
+ */
+function ExecutionProfile(options) {
+  options = options || utils.emptyObject;
+  /**
+   * Consistency level
+   * @type {Number}
+   */
+  this.consistency = options.consistency;
+  /**
+   * Load-balancing policy
+   * @type {LoadBalancingPolicy}
+   */
+  this.loadBalancing = options.loadBalancing;
+  /**
+   * Client read timeout.
+   * @type {Number}
+   */
+  this.readTimeout = options.readTimeout;
+  /**
+   * Retry policy.
+   * @type {RetryPolicy}
+   */
+  this.retry = options.retry;
+  /**
+   * Serial consistency level.
+   * @type {Number}
+   */
+  this.serialConsistency = options.serialConsistency;
+}
+
+/**
+ * @param {ClientOptions} options
+ * @constructor
+ * @ignore
+ */
+function ProfileManager(options) {
+  this.profiles = options.profiles || {};
+  this.setDefault(options);
+  // use an array of profiles to iterate more efficiently
+  this.profilesArray = Object.keys(this.profiles).map(function (name) {
+    // set required properties
+    var p = this.profiles[name];
+    p.loadBalancing = p.loadBalancing || this._defaultProfile.loadBalancing;
```

Comment:
One thing I bumped into that I couldn't wrap my head around at first was that it is probably not a good idea to share ExecutionProfile instances between Client instances.  Just like you wouldn't share LoadBalancingPolicies because it has host-specific details.  We should document profiles should not be shared between client instances.  Alternatively we could decorate ExecutionProfiles as they are evaluated with some kind of property indicating that they are already tied to a Cluster or something to that effect.